core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/latency/VMOperationRule.java

```diff
@@ -92,36 +98,112 @@ private Result evaluate(IItemCollection items, IQuantity warningLimit) {
-				.getAggregate(Aggregators.itemWithMax(JfrAttributes.DURATION));
 
-		if (longestEvent == null) {
+		findLongestEventInfo(items.apply(JdkFilters.VM_OPERATIONS_BLOCKING_OR_SAFEPOINT));
+		if (startingEvent == null) {
 			String zeroDuration = UnitLookup.SECOND.quantity(0).displayUsing(IDisplayable.AUTO);
 			return new Result(this, 0,
 					MessageFormat.format(Messages.getString(Messages.VMOperationRuleFactory_TEXT_OK), zeroDuration),
 					null, JdkQueries.VM_OPERATIONS);
 		}
-		IQuantity longestDuration = RulesToolkit.getValue(longestEvent, JfrAttributes.DURATION);
-		String timeStr = RulesToolkit.getValue(longestEvent, JfrAttributes.START_TIME).displayUsing(IDisplayable.AUTO);
+		String timeStr = getStartTime(startingEvent).displayUsing(IDisplayable.AUTO);
 		String peakDuration = longestDuration.displayUsing(IDisplayable.AUTO);
+		String operation = getOperation(startingEvent);
+		IMCThread caller = getCaller(startingEvent);
 		double score = RulesToolkit.mapExp100(longestDuration.doubleValueIn(UnitLookup.SECOND),
 				infoLimit.doubleValueIn(UnitLookup.SECOND), warningLimit.doubleValueIn(UnitLookup.SECOND));
 
+		boolean isCombinedDuration = getDuration(startingEvent).compareTo(longestDuration) != 0;
+		longestDuration = null;
+		startingEvent = null;
 		if (Severity.get(score) == Severity.WARNING || Severity.get(score) == Severity.INFO) {
-			String operation = RulesToolkit.getValue(longestEvent, JdkAttributes.OPERATION);
-			IMCThread caller = RulesToolkit.getValue(longestEvent, JdkAttributes.CALLER);
-			return new Result(this, score,
-					MessageFormat.format(Messages.getString(Messages.VMOperationRuleFactory_TEXT_WARN), peakDuration),
-					MessageFormat.format(Messages.getString(Messages.VMOperationRuleFactory_TEXT_WARN_LONG),
-							peakDuration, operation, Encode.forHtml(caller.getThreadName()), timeStr),
+			String longMessage = isCombinedDuration ? Messages.VMOperationRuleFactory_TEXT_WARN_LONG_COMBINED_DURATION
+					: Messages.VMOperationRuleFactory_TEXT_WARN_LONG;
+			String shortMessage = isCombinedDuration ? Messages.VMOperationRuleFactory_TEXT_WARN_COMBINED_DURATION
+					: Messages.VMOperationRuleFactory_TEXT_WARN;
+			return new Result(this, score, MessageFormat.format(Messages.getString(shortMessage), peakDuration),
+					MessageFormat.format(Messages.getString(longMessage), peakDuration, operation, caller, timeStr),
 					JdkQueries.VM_OPERATIONS_BLOCKING);
 		}
-		return new Result(this, score,
-				MessageFormat.format(Messages.getString(Messages.VMOperationRuleFactory_TEXT_OK), peakDuration), null,
+		String shortMessage = isCombinedDuration ? Messages.VMOperationRuleFactory_TEXT_OK_COMBINED_DURATION
+				: Messages.VMOperationRuleFactory_TEXT_OK;
+		return new Result(this, score, MessageFormat.format(Messages.getString(shortMessage), peakDuration), null,
 				JdkQueries.FILE_READ);
 	}
 
+	private void findLongestEventInfo(IItemCollection items) {
```

Comment:
For this method it would be much better to use the established pattern for iterating through IItemCollections:
`for (IItemIterable itemIterable : items) {
  // get your accessors here
  IMemberAccessor<IQuantity, IItem> startTime = JfrAttributes.START_TIME.getAccessor(itemIterable.getType());
  for (IItem item : itemIterable) {
    IQuantity time = startTime.getMember(item);
  }
}`

Each IItemIterable represents one event lane of one event type of disjoint events, sorted by timestamp.